### DISCUSSION TITLE
GENERATING INPUT FILE-Generating complex Geometry (trelis)

### DISCUSSION THREAD

**Discussion by japresa** [idaholab/moose#19268] [Q&A General] [answered]
Hello all, 
planning another project with moose, and I am having a meshed input file with a bunch of different elements. I previously used Trelis software to generate the simulation-meshed-geometry-exodusll-input-files, and made them by hand. However, this next project has too many elements to insert by hand.

Do you have any recommendations on how to effectively generate an exodus-input-file for moose, that can increase the speed for bigger projects? 

I'm thinking of writing a higher-level-programming-script to interact with the Trelis GUI and automate the "create object" and "move to this location," processes. If there's any other way that any of you have discovered lease let me know.

Anything helps. Thank you!

**Comment by loganharbour** [comment]
Can you describe the geometry? Or the process you are trying to automate?

**Reply by japresa** [reply]
![image](https://user-images.githubusercontent.com/79496775/139619725-d986d444-93d4-48fa-870e-9d44b71c549c.png)

Yes.
I'm modeling a 3D core/(can) with a bunch of cylinder-holes in it, like in this picture. Its for heat pipes and heat sources that will be inserted. 

--------------------------------------
The steps that I'm trying to automate are the following:

cut out cylinder-holes from the core/can at locationX from a CSV file I generated
reinsert cylinders(heat sources) where the holes were at 

Add all cylinders(heat sources) to a common-block-object 
Define side-sets for where the cylinders(heat sources) touch the core/can

**Reply by GiudGiud** [reply]
Are you meshing the whole domain or are you meshing only the cylindrical holes (are they cylinders? or hexagons like in the picture?)

Step 3 and 4 of your process seem like they can be done in MOOSE using mesh generators. Let s think about it for the rest

**Reply by japresa** [reply]
I'm meshing the whole domain. It is a pentagon-shaped-3dCore with a smaller cylinders(not pentagons) inserted in it. 

That's a relief that steps 3/4 are possible by using MOOSE generators, that's one of the biggest hurdle because step 3 alone would require 1000+ sidesets. Just for clarification, are you implying the following: importing an unmeshed-geometry with no blocks or side sets, and then defining those using the MeshGenerator linked below? I'll take a closer look at the MeshGenerator tomorrow morning.

**Reply by GiudGiud** [reply]
You can add blocks and sidesets in MOOSE with parsed expressions. If you have 1000s of sidesets to add you will want to program a new mesh generator to do it

**Comment by GiudGiud** [comment]
Mesh generation can also be done in MOOSE in some capacity, though mostly for simple geometries, and lattice-based geometries. See the mesh generator system https://mooseframework.inl.gov/source/meshgenerators/MeshGenerator.html

Guillaume

**Comment by GregVernon** [comment, marked as answer]
> I'm thinking of writing a higher-level-programming-script to interact with the Trelis GUI and automate the "create object" and "move to this location," processes. If there's any other way that any of you have discovered lease let me know.

@japresa -- Trelis is now [Coreform Cubit](https://coreform.com/products/coreform-cubit/) and it has a [Python API](https://coreform.com/cubit_help/cubithelp.htm#t=appendix%2Fpython%2Fcubit_python_interface.htm).  This API was also available in Trelis.  We (Coreform) recorded a webinar earlier this year on using the Python API: https://www.youtube.com/watch?v=2TCPPO8z5VA

**Reply by GiudGiud** [reply]
@japresa I think this is the preferred way. I did not remember that trelis was the old name for coreform cubit... We actually use the python API for cubit all the time.

I have examples online on the virtual test bed, though it s for 2D RZ meshes
https://github.com/idaholab/virtual_test_bed/tree/devel/msr/msfr/mesh
https://github.com/idaholab/virtual_test_bed/blob/devel/pbfhr/meshes/generate_core_with_reflectors.py
